unixfs/tar/writer.go

```diff
@@ -21,79 +21,100 @@ import (
 // TODO: does this need to be configurable?
 var DefaultBufSize = 1048576
 
-func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, compression int) (io.Reader, error) {
+// DagArchive is equivalent to `ipfs getdag $hash | maybe_tar | maybe_gzip`
+func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, useTar bool, compression int) (io.Reader, error) {
 
 	_, filename := path.Split(name)
 
 	// need to connect a writer to a reader
 	piper, pipew := io.Pipe()
 
-	// use a buffered writer to parallelize task
-	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
-
-	// construct the tar writer
-	w, err := NewWriter(bufw, dag, compression)
+	// compression determines whether to use gzip compression.
+	gzw, err := gzip.NewWriterLevel(pipew, compression)
 	if err != nil {
 		return nil, err
 	}
 
+	w := &Writer{
+		Dag: dag,
+		GzW: gzw,
+		ctx: ctx,
```

Comment:
yeah i debating putting the ctx here too. I opted for the function as it is only needed there-- otherwise the reader can run indefinitely.